tests/rptest/tests/retention_policy_test.py

```diff
@@ -575,3 +576,134 @@ def cloud_log_size() -> int:
+
+        # A fictional artificial timestamp base in milliseconds
+        future_timestamp = (int(time.time()) + 24 * 3600) * 1000
+
+        # Produce a run of messages with CreateTime-style timestamps, each
+        # record having a timestamp 1ms greater than the last.
+        msg_size = 14000
+        segments_count = 10
+        msg_count = (self.segment_size // msg_size) * segments_count
+
+        if mixed_timestamps:
+            valid_records = (self.segment_size // msg_size) // 2
+
+            # Write enough valid-timestamped records that one should appear in the index
+            producer = KgoVerifierProducer(context=self.test_context,
+                                           redpanda=self.redpanda,
+                                           topic=self.topic,
+                                           msg_size=msg_size,
+                                           msg_count=valid_records,
+                                           batch_max_bytes=msg_size * 2)
+            producer.start()
+            producer.wait()
+            producer.free()
+
+            # Write the rest of the messages with invalid timestamps
+            producer = KgoVerifierProducer(context=self.test_context,
+                                           redpanda=self.redpanda,
+                                           topic=self.topic,
+                                           msg_size=msg_size,
+                                           msg_count=msg_count - valid_records,
+                                           fake_timestamp_ms=future_timestamp,
+                                           batch_max_bytes=msg_size * 2)
+            producer.start()
+            producer.wait()
+        else:
+            # Write msg_count messages with timestamps in the future
+            producer = KgoVerifierProducer(
+                context=self.test_context,
+                redpanda=self.redpanda,
+                topic=self.topic,
+                msg_size=msg_size,
+                msg_count=(self.segment_size // msg_size) * segments_count,
+                fake_timestamp_ms=future_timestamp,
+                batch_max_bytes=msg_size * 2)
+            producer.start()
+            producer.wait()
+
+        # We should have written the expected number of segments, and nothing can
+        # have been gc'd yet because all the segments have max timestmap in the future
+        segs = self.redpanda.node_storage(self.redpanda.nodes[0]).segments(
+            "kafka", self.topic, 0)
+        self.logger.debug(f"Segments after write: {segs}")
+        assert len(segs) >= segments_count
+
+        # Give retention code some time to kick in: we are expecting that it does not
+        # remove anything because the timestamps are too far in the future.
+        with self.redpanda.monitor_log(self.redpanda.nodes[0]) as mon:
+            # Time period much larger than what we set log_compaction_interval_ms to
+            sleep(10)
```

Comment:
Is this sleep actually necessary? The wait until below implies that retention has kicked in.

Reply:
Mostly not, but I wanted to have a better chance of the housekeeping having entirely completed, rather than proceeding as soon as it has seen one segment that emits the bogus timestamp message.